tests/shell/tests/01-run.py

```diff
@@ -22,9 +22,9 @@
 )
 
 EXPECTED_PS = (
-    '\tpid | state    Q | pri',
-    '\t  1 | pending  Q |  15',
-    '\t  2 | running  Q |   7'
+    r'\s+pid | state    Q | pri',
+    r'\s+1 | pending  Q |\s+\d+',
```

Comment:
Is this related to ESP32? `SCHED_PRIO_LEVELS` seems to be not architecture specific but a common variable that might change from application to application as well.

Reply:
This is not ESP32 related. I'll cut out the commit into another PR.

Reply:
the esp Makefile states ```# if any WiFi interface is used,  the number of priority levels has to be 32```.

Do you remember the reason why?

A couple of tests fail because the priorities change with ```SCHED_PRIO_LEVELS=32```.

Reply:
Yes, number of priority levels is simply to small if `esp_wifi` is used. If `esp_wifi` is used there are 3 additional high-priority threads:

- WiFi hardware handler `wifi` with priority 1
- WiFi event loop `wifi-event-loop` with priority 4 which is defined by the ESP32 SDK.
- WiFi netdev `netdev-esp-wifi` with priority 10 which is simply the priority defined by `GNRC_NETIF_PRIO = (THREAD_PRIORITY_MAIN - 5)`.

```
> ps
	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   2048 (  400) | 0x3ffb0220 | 0x3ffb0a20
	  1 | wifi-event-loop      | bl rx    _ |   4 |   2100 ( 1020) | 0x3ffaee18 | 0x3ffaf430 
	  2 | idle                 | pending  Q |  31 |   2048 (  448) | 0x3ffb4b80 | 0x3ffb51e0 
	  3 | main                 | running  Q |  15 |   3072 ( 1616) | 0x3ffb5380 | 0x3ffb5d00 
	  4 | pktdump              | bl rx    _ |  14 |   3072 (  560) | 0x3ffb9da0 | 0x3ffba770 
	  5 | ipv6                 | bl rx    _ |  12 |   2048 (  752) | 0x3ffb7890 | 0x3ffb7e40 
	  6 | udp                  | bl rx    _ |  13 |   2048 (  576) | 0x3ffbb3f0 | 0x3ffbb9b0 
	  7 | wifi                 | bl rx    _ |   1 |   3636 ( 1848) | 0x3ffbe004 | 0x3ffbec10 
	  8 | netdev-esp-wifi      | bl rx    _ |  10 |   2048 (  848) | 0x3ffb6490 | 0x3ffb6a10 
	  9 | RPL                  | bl rx    _ |  13 |   2048 (  492) | 0x3ffbabec | 0x3ffbb200 
	    | SUM                  |            |     |  24168 ( 8560)
```
If you would use the default number of priority levels, `THREAD_PRIORITY_MAIN` would be 7 and `GNRC_NETIF_PRIO` would be 2. This could lead to a deadlock situations because upper layer  `netdev-esp-wifi` has higher priority than lower layer thread `wifi-event-loop` which is dealing with events from hardware driver thread `wifi`. At least, there will be instabilities if thread `wifi-event-loop` can't handle events from WiFi hardware just in time.

Following log shows the priorities with default number of priority levels.
```
> ps
	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   2048 (  400) | 0x3ffb0220 | 0x3ffb0a20
	  1 | wifi-event-loop      | bl rx    _ |   4 |   2100 ( 1020) | 0x3ffaee18 | 0x3ffaf430 
	  2 | idle                 | pending  Q |  15 |   2048 (  416) | 0x3ffb4b40 | 0x3ffb51a0 
	  3 | main                 | running  Q |   7 |   3072 ( 1616) | 0x3ffb5340 | 0x3ffb5cc0 
	  4 | pktdump              | bl rx    _ |   6 |   3072 (  560) | 0x3ffb9d60 | 0x3ffba730 
	  5 | ipv6                 | bl rx    _ |   4 |   2048 (  752) | 0x3ffb7850 | 0x3ffb7e00 
	  6 | udp                  | bl rx    _ |   5 |   2048 (  576) | 0x3ffbb3b0 | 0x3ffbb970 
	  7 | wifi                 | bl rx    _ |   1 |   3636 ( 1864) | 0x3ffbdfc4 | 0x3ffbebd0 
	  8 | netdev-esp-wifi      | bl rx    _ |   2 |   2048 (  848) | 0x3ffb6450 | 0x3ffb69d0 
	  9 | RPL                  | bl rx    _ |   5 |   2048 (  492) | 0x3ffbabac | 0x3ffbb1c0 
	    | SUM                  |            |     |  24168 ( 8544)
```
It seems that there are too many threads with same priority. I don't know remember where, but I have read anywhere in Wiki or the online doc that each thread should have it's own priority since threads with same priority don't lead to a context switch as long as both are in state pending.

Reply:
BTW, I have seen that there is still an inconsistency in `cpu/esp32/Makefile.include`. `SCHED_PRIO_LEVELS` is always set to 32.

```make
CFLAGS  += -DSCHED_PRIO_LEVELS=32
...
# if any WiFi interface is used,  the number of priority levels has to be 32
ifneq (,$(filter esp_wifi_any,$(USEMODULE)))
    CFLAGS += -DSCHED_PRIO_LEVELS=32
endif
```

Reply:
> I don't know remember where, but I have read anywhere in Wiki or the online doc that each thread should have it's own priority since threads with same priority don't lead to a context switch as long as both are in state pending.

I know the statement, it is a bit misleading. If two (or more) threads share a priority and one doesn't either yield or block, the other thread(s) can starve, same as lower prioritized threads. Other than that, having two threads with the same priority is fine.

Reply:
> If `esp_wifi` is used there are 3 additional high-priority threads:
> 
>     * WiFi hardware handler `wifi` with priority 1
> 
>     * WiFi event loop `wifi-event-loop` with priority 4 which is defined by the ESP32 SDK.
> 
>     * WiFi netdev `netdev-esp-wifi` with priority 10 which is simply the priority defined by `GNRC_NETIF_PRIO = (THREAD_PRIORITY_MAIN - 5)`.

So they have to be ordered? would this be fine:

wifi 0
wifi-event-loop 1
netdev-esp-wifi 2 (default)

?

Reply:
It should work. Unfortunately, the priority of the `wifi-event-loop` is not defined in the RIOT code, but in the SDK.

What couple of tests fail? I would like to take a look why.

IMHO, it is a more general problem. If there is a variable that allows to define the number of priorities, I would like to be able to use it. Not using it because some tests fail doesn't seem to be the right solution.

Reply:
> What couple of tests fail? I would like to take a look why.

Take a look at the CI output: https://ci.riot-os.org/RIOT-OS/RIOT/11449/3c5e7c9b76d0f2da5629c9298ae3234b8f75c2c1/output.html

I only skimmed the outputs. Some are expecting default priorities, some fail for other reasons.

> IMHO, it is a more general problem. If there is a variable that allows to define the number of priorities, I would like to be able to use it. Not using it because some tests fail doesn't seem to be the right solution.

Sure, but rearranging one architecture's configuration a bit (to make it more like all the others) is IMO preferable than making the tests less exact (by accepting *any* priority instead of what it should be in the default configuration.

Reply:
> It should work. Unfortunately, the priority of the `wifi-event-loop` is not defined in the RIOT code, but in the SDK.

It is defined here, right?

```
[kaspar@ng riot (master)]$ git grep -A1 -n -- wifi-event-loop
cpu/esp32/vendor/esp-idf/esp32/event_loop.c:120:    xTaskCreatePinnedToCore(esp_event_loop_task, "wifi-event-loop",
cpu/esp32/vendor/esp-idf/esp32/event_loop.c-121-            ESP_TASKD_EVENT_STACK, NULL, ESP_TASKD_EVENT_PRIO, NULL, 0);
[kaspar@ng riot (master)]$
```

I think you were referring to the "wifi" hardware handler thread. Cannot find it's definition within the RIOT codebase.

Would wifi in 1, wifi-event-loop on 2 and the netif thread on 3 work?

Reply:
> It is defined here, right?
> 
> ```
> [kaspar@ng riot (master)]$ git grep -A1 -n -- wifi-event-loop
> cpu/esp32/vendor/esp-idf/esp32/event_loop.c:120:    xTaskCreatePinnedToCore(esp_event_loop_task, "wifi-event-loop",
> cpu/esp32/vendor/esp-idf/esp32/event_loop.c-121-            ESP_TASKD_EVENT_STACK, NULL, ESP_TASKD_EVENT_PRIO, NULL, 0);
> [kaspar@ng riot (master)]$
> ```

Yes, nut it is vendor code, I tried to modify as less as possible to make future updates as easy as possible. Sure, we can override the definition here.

> I think you were referring to the "wifi" hardware handler thread. Cannot find it's definition within the RIOT codebase.

No, I was refering the vendor code above.
 
> Would wifi in 1, wifi-event-loop on 2 and the netif thread on 3 work?

It should work but I don't know what the side effects are if we simply increase `GNRC_NETIF_PRIO`.

Reply:
I took a look to the tests that failed. All of them are tests that don't require module `esp_wifi`. So, I would prefer to reactivate the conditional setting of `SCHED_PRIO_LEVELS`.

```diff
 CFLAGS  += -DSCHED_PRIO_LEVELS=32
-CFLAGS  += -DSCHED_PRIO_LEVELS=32
...
# if any WiFi interface is used,  the number of priority levels has to be 32
ifneq (,$(filter esp_wifi_any,$(USEMODULE)))
    CFLAGS += -DSCHED_PRIO_LEVELS=32
endif

This should help the tests which insist on certain PIDs and will work for networking applications as before.

Reply:
> So, I would prefer to reactivate the conditional setting of `SCHED_PRIO_LEVELS`.

ok!

Reply:
This did the trick for most tests.

Reply:
The `ps` regex is currently wrong, as `|` is a regex character. It should be escaped to have a proper test.

Reply:
These changes are not necessary any longer. After changing the thread priority levels in PR #12753 everything is working without changes.